File: testdata(temporary)/extract_colors.js
const sharp = require("sharp");
const https = require("https");
const fs = require("fs");
const path = require("path");

const DATA = path.join(__dirname, "allbirds_products.json");
const OUT = path.join(__dirname, "allbirds_products.json");
const products = JSON.parse(fs.readFileSync(DATA, "utf-8"));

async function getDominantColor(imgUrl) {
  const buf = await new Promise((resolve, reject) => {
    https.get(imgUrl, (res) => {
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve(Buffer.concat(chunks)));
    }).on("error", reject);
  });

  const { data } = await sharp(buf)
    .resize(50, 50, { fit: "fill" })
    .raw()
    .toBuffer({ resolveWithObject: true });

  const buckets = {};
  for (let i = 0; i < data.length; i += 3) {
    const r = Math.min(255, Math.round(data[i] / 32) * 32);
    const g = Math.min(255, Math.round(data[i + 1] / 32) * 32);
    const b = Math.min(255, Math.round(data[i + 2] / 32) * 32);
    const key = `${r},${g},${b}`;
    if (r + g + b > 650 || r + g + b < 80) continue;
    buckets[key] = (buckets[key] || 0) + 1;
  }

  const entries = Object.entries(buckets).sort((a, b) => b[1] - a[1]);
  if (!entries.length) return "#c8c8c8";
  const [r, g, b] = entries[0][0].split(",").map(n => +n);
  return `#${[r, g, b].map(n => n.toString(16).padStart(2, "0")).join("")}`;
}

(async () => {
  const targets = products.filter(p => p.tags?.includes("collection:apr26"));
  console.log(`Processing ${targets.length} products...`);

  for (let i = 0; i < targets.length; i++) {
    const p = targets[i];
    const imgUrl = `https:${p.images?.[0]?.preview_image?.src}`;
    if (!imgUrl || imgUrl === "https:undefined") {
      p.color = "#c8c8c8";
    } else {
      try {
        p.color = await getDominantColor(imgUrl);
      } catch {
        p.color = "#c8c8c8";
      }
    }
    process.stdout.write(`\r${i + 1}/${targets.length} - ${p.color}`);
  }

  fs.writeFileSync(OUT, JSON.stringify(products, null, 2));
  console.log(`\nDone. Written to ${OUT}`);
})();

File: testdata(temporary)/extract_colors.py
import json, urllib.request, os
from PIL import Image
from io import BytesIO
from collections import Counter

DATA = os.path.join(os.path.dirname(__file__), "allbirds_products.json")
OUT = os.path.join(os.path.dirname(__file__), "allbirds_products.json")

with open(DATA) as f:
    products = json.load(f)

targets = [p for p in products if "collection:apr26" in p.get("tags", [])]
print(f"Processing {len(targets)} products...")

for idx, p in enumerate(targets):
    src = p.get("images", [{}])[0].get("preview_image", {}).get("src")
    if not src:
        p["colorcode"] = "#c8c8c8"
        continue
    url = f"https:{src}"

    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
        data = urllib.request.urlopen(req, timeout=15).read()
        img = Image.open(BytesIO(data)).convert("RGB")
        img = img.resize((50, 50), Image.BILINEAR)
        pixels = list(img.getdata())

        buckets = Counter()
        for r, g, b in pixels:
            r = round(r / 32) * 32
            g = round(g / 32) * 32
            b = round(b / 32) * 32
            if r + g + b > 650 or r + g + b < 80:
                continue
            key = f"{r},{g},{b}"
            buckets[key] += 1

        if not buckets:
            p["colorcode"] = "#c8c8c8"
        else:
            top = buckets.most_common(1)[0][0]
            parts = [int(x) for x in top.split(",")]
            p["colorcode"] = "#" + "".join(f"{min(255, v) & 255:02x}" for v in parts)
    except Exception as e:
        print(f"\nError on {idx}: {e}")
        p["colorcode"] = "#c8c8c8"

    if (idx + 1) % 10 == 0:
        print(f"  {idx + 1}/{len(targets)}")

with open(OUT, "w") as f:
    json.dump(products, f, indent=2)

# Also write JS module
JS = os.path.join(os.path.dirname(__file__), "allbirds_products.js")
with open(JS, "w") as f:
    f.write("export const data = ")
    json.dump(products, f, indent=2)
    f.write(";\n")

print(f"Done. {sum(1 for p in products if 'colorcode' in p)} products have colorcode")
